test/13_arraycommon.jl

```diff
@@ -1827,4 +1827,89 @@ end
     end
 end
 
+using JSON3
+using StructTypes
+
+struct MyCustomTypeMissing
+    id::Vector{Int}
+    var::CategoricalVector{Union{Missing,String}}
+end
+StructTypes.StructType(::Type{<:MyCustomTypeMissing}) = StructTypes.Struct()
+
+struct MyCustomType
+    id::Vector{Int}
+    var::CategoricalVector{String}
+end
+StructTypes.StructType(::Type{<:MyCustomType}) = StructTypes.Struct()
+
+@testset "Reading CategoricalVector objects using JSON3" begin
+    x = CategoricalArray(["x","y","z","y","y","z"])
+    str = JSON3.write(x)
+    readx = JSON3.read(str, CategoricalArray)
+    @test readx == x
+    @test levels(readx) == levels(x)
+    @test readx isa CategoricalArray
+
+    x = CategoricalArray([missing,"y","z","y",missing,"z","x"])
+    str = JSON3.write(x)
+
+    readx = JSON3.read(str, CategoricalVector)
+    @test x ≅ readx
+    @test all(sort(levels(readx)) .== sort(levels(x)))
```

Comment:
```suggestion
    @test sort(levels(readx)) == sort(levels(x))
```

BTW, shouldn't the test be a bit stricter about the expected order?
```suggestion
    @test levels(readx) == sort(levels(x))
```

Reply:
These need to use all because the levels array for `readx` is of type `Array{Any}`; t

The above discussion with quinnj where I realize I am maybe making this more complicated than I need to is about this issue. To make the array something more narrow than `Any` I would need to either change how JSON3 works or change `categoricalgeneral` to use that messy hacky bit where I check the runtime types of each element in the array.

Reply:
`==` doesn't take into account the eltype, so it should work nevertheless. But it would be nice to test the full type of `readx` for each case as it matters a lot for performance.

Is the problem that JSON3 creates `Array{Any}` objects rather than `Array{Union{T, Missing}` objects? If so, you can call `identity.(A)` to let `collect` compute the narrowest eltype. Use `ifelse.(A .=== nothing, missing, A)` when replacing `nothing` to avoid allocating two copies.

Reply:
Okay, I've switched to `==`.

And thanks for the type about `identity`: I made that change as well.